### DISCUSSION TITLE
Add support for sending a file

### DISCUSSION THREAD

**Discussion by prabirshrestha** [trillium-rs/trillium#112] [Ideas]
I have been working on porting a [tide app to use trillium](https://github.com/prabirshrestha/rblog/pull/8)  and noticed that there is no way to send a single file.  Closest I found was `StaticFileHandler`. 

I'm looking for an alternative for `Body::from_file()`.

```rust
 let res = Response::builder(StatusCode::Ok)
                .header("cache-control", "max-age=31536000") // 1 year as second
                .body(Body::from_file(&attachment.get_path()).await?)
                .build();
```

**Comment by jbr** [comment]
Converted to a discussion, as this is potentially a new handler / crate, specifically if the static file handler doesn't work for you.

**Reply by prabirshrestha** [reply]
For now I created a small util function.

```rust
/// requires async-fs (smol) crate
pub async fn body_from_path<P: AsRef<std::path::Path>>(path: P) -> std::io::Result<Body> {
    use async_fs::{self as fs, File};
    let file = File::open(&path).await?;
    let len = fs::metadata(&path).await?.len();
    Ok(Body::new_streaming(file, Some(len)))
}

/// requires mime_guess crate
pub async fn body_from_path_with_mime<P: AsRef<std::path::Path>>(
    path: P,
) -> std::io::Result<(Body, String)> {
    Ok((
        body_from_path(&path).await?,
        mime_guess::from_path(&path)
            .first_or_octet_stream()
            .to_string(),
    ))
}
```

Then one can use it as

```rust
    let (body, content_type) =
        conn_try!(body_from_path_with_mime("/tmp/file").await, conn);

    conn.with_header(KnownHeaderName::CacheControl, "max-age=31536000") // 1 year as a second
        .with_header(KnownHeaderName::ContentType, content_type)
        .ok(body)
```

The only problem is I get a random number in  response body. Not sure what 25 is and seems to be truncating text.

```
25
This is a sample welcome attachme
```

Original file content is `This is a sample welcome attachment.`. https://github.com/prabirshrestha/rblog/blob/7daff9b9b52f018ea4d67dfca061db84377aca63/posts/welcome/attachment.txt

**Reply by prabirshrestha** [reply]
Seems like if I use `Body::new_streaming(file, None))` instead of `Body::new_streaming(file, Some(len))` it seems to work properly. What is len when using `new_streaming()`?

**Reply by jbr** [reply]
This is fixed on main, will release today. Your code is correct.

**Comment by prabirshrestha** [comment]
What is the recommended way to create an `async Conn` extension trait. I would like to accomplish `|conn: Conn| conn.send_file("file.txt")` but the following would work since the caller would need to await as well as `Conn` is not a `Send`.

```rust
    use async_fs::{self as fs, File};

    #[trillium::async_trait]
    pub trait SendFileConnExt {
        async fn send_file<P: AsRef<Path>>(mut self, path: P) -> Self;
    }

    #[trillium::async_trait]
    impl SendFileConnExt for Conn {
        async fn send_file<P: AsRef<Path>>(mut self, path: P) -> Self {
            let file = conn_try!(File::open(&path).await, self);
            let metadata = conn_try!(fs::metadata(&path).await, self);
            let len = metadata.len();
            self.with_body(Body::new_streaming(file, Some(len)))
                .with_status(200)
        }
    }

```

**Reply by prabirshrestha** [reply]
I have now published the [trillium-send-file](https://github.com/prabirshrestha/trillium-send-file) crate. You can use `cargo add trillium-send-file`. Currently only supports smol and async-std runtimes.

Usage:

```rust
use trillium::Conn;
use trillium_send_file::SendFileConnExt;

fn main() {
    trillium_smol::run(|conn: Conn| async move {
        conn.send_file("/tmp/file.txt".into()).await
    });
}
```

**Reply by jbr** [reply]
Cool, nice work! I've been working on some changes to trillium-static that overlap with this, but it's good for the ecosystem to have several options

**Reply by prabirshrestha** [reply]
Just saw your PR, seems like it has almost all the features I wanted to implement in my send file. I was hoping to work on etag support in the upcoming weeks but seems like you have already implemented. I will move to your library.

Do you also plan to support [range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) so clients can support resuming downloads or downloading in chunks? Wondering if trillium needs `Body::new_seekable_streaming()` that allows `AsyncSeek`.

Unrelated to this thread but what are your plans on supporting multipart uploads?

**Reply by jbr** [reply]
(wrote this as a distinct comment and lost the first version when I moved it to a reply)

Yes to both of these, it's just a matter of order. Range requests (and an associated feature, http trailers for content-based etags on streaming bodies) require changes to the `trillium-http` crate, which means I probably will wait a bit while I think it through and sketch out several different approaches before publishing.

Multipart I had figured was low priority because I wasn't sure if anyone actually used html forms directly these days without js, but probably could be convinced to prioritize it. Unlike range requests, it wouldn't require any changes to the `trillium-http` or `trillium` crates, which makes it easier to get it wrong and iterate

**Reply by prabirshrestha** [reply]
Not in hurry for range requests yet so this can come later.

Multipart would be needed if one want to implement it using pure html forms and server side without javascript.

If we want to pass additional metadata about the file I would also want to use multipart even if it is via ajax. While one could use a rest api and then add headers I would like to have a a url such as `/upload` and `filename` as multipart and the actual file content as the second part of the multipart content.
